### PR TITLE
Dynamically assign DNS IP using cni-env

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
### Summary
Earlier we used to assign DNS IP statically. But now we use the CIDR in cni-env to assign a valid IP from that range removing hard-coded IP. 
We add new utilities in the utils.sh to calculate and test a valid IP from the cidr. Not using external utils like `ipcalc`, as they may not be present in every system.

### CODE REVIEW

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Is it not better to simply deploy the service without a fixed IP address, and then query KubeAPIserver with the address?

I cannot see the reason to add all this logic just to pick a random IP of a valid range, when Kubernetes will do that for us either way.

**Comment by sachinkumarsingh092** [comment]
We now assign DNSIP after deploying the service and assign arguments in kubelet based on the Cluster IP assigned to the DNS service. We also remove DNSIPARGS and assign IP to the DNS service dynamically only.

**Review by neoaggelos — CHANGES_REQUESTED** [review]
This works much better. However, setting the DNS IP address as a second argument must still be supported.

You can support both by setting `clusterIP` if $2 is provided, otherwise leave it as an empty value.

**Comment by sachinkumarsingh092** [comment]
But the DNSIP env variable cannot be an empty value inside the manifest. Otherwise we will get errors like:
```
The Service "kube-dns" is invalid: spec.clusterIPs[0]: Invalid value: "$DNSIP": must be a valid IP address, (e.g. 10.9.8.7 or 2001:db8::ffff)
```

**Comment by neoaggelos** [comment]
Yes, this is because `$DNSIP` is not replaced in the manifest, and is rejected because of this. Try the following:

```yaml
# works
apiVersion: v1
kind: Service
metadata:
  name: svc-1
spec:
  clusterIP: 
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
---
# does not work
apiVersion: v1
kind: Service
metadata:
  name: svc-2
spec:
  clusterIP: "$DNSIP"
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
```

File: addons/dns/coredns.yaml [hunk 1]
```diff
@@ -146,7 +146,7 @@ metadata:
 spec:
   selector:
     k8s-app: kube-dns
-  clusterIP: $DNSIP
+  clusterIP:
```
Comment (neoaggelos) [inline addons/dns/coredns.yaml:149]: This does not look right.

**Review by neoaggelos — CHANGES_REQUESTED** [review]

File: addons/dns/enable [hunk 2]
```diff
@@ -47,13 +47,12 @@ else
...
-DNSIPARG=$2
-if [[ -n "$DNSIPARG" ]]; then
-  DNSIP="$DNSIPARG";
-  echo "Will use ${DNSIPARG} for dns service ip address"
-  echo "Make sure this address is in the service cidr!"
+DNSIP="$2"
+if [ x"${DNSIP}" == "x" ]; then
+  # (QOL): If not using a custom service CIDR, then use the well-known address 10.152.183.10
+  if cat "${SNAP_DATA}/args/kube-apiserver" | grep -q -- '^--service-cluster-ip-range=10.152.183.0/24' 2>/dev/null; then
+    DNSIP="10.152.183.10"
```
Comment (sachinkumarsingh092) [inline addons/dns/enable:54]: Are we not trying to remove any kind of hardcodings when assigning IPs?

Comment (neoaggelos) [reply addons/dns/enable:54]: In the default situation where no custom CIDR is set, this saves us from an unnecessary restart of the kubelite service (which would add 1+ minute of initial setup until all services have acquired their lease.

**Review by neoaggelos — APPROVED** [review]

**Pull request merged**